core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/RunningStats.java

```diff
@@ -0,0 +1,312 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.metrics.stats.multifield;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Descriptive stats gathered per shard. Coordinating node computes final correlation and covariance stats
+ * based on these descriptive stats
+ *
+ * @internal
+ */
+public class RunningStats implements Streamable, Cloneable {
+    /** count of observations (same number of observations per field) */
+    protected long docCount;
+    /** per field sum of observations */
+    protected HashMap<String, Double> fieldSum;
+    /** counts */
+    protected HashMap<String, Long> counts;
+    /** mean values */
+    protected HashMap<String, Double> means;
+    /** variance values */
+    protected HashMap<String, Double> variances;
+    /** skewness values */
+    protected HashMap<String, Double> skewness;
+    /** kurtosis values */
+    protected HashMap<String, Double> kurtosis;
+    /** covariance values */
+    protected HashMap<String, HashMap<String, Double>> covariances;
+
+    /** Ctor to create an instance of running statistics */
+    public RunningStats() {
+        docCount = 0;
+        counts = new HashMap<>();
+        fieldSum = new HashMap<>();
+        means = new HashMap<>();
+        skewness = new HashMap<>();
+        kurtosis = new HashMap<>();
+        covariances = new HashMap<>();
+        variances = new HashMap<>();
+    }
+
+    /** adds a documents fields to the running statistics **/
+    public void add(Map<String, Double> docFields) {
+        // increment count
+        docCount++;
+
+        // update total, mean, and variance
+        String fieldName;
+        double fieldValue;
+        final HashMap<String, Double> delta = new HashMap<>();
+        for (Map.Entry<String, Double> field : docFields.entrySet()) {
+            fieldName = field.getKey();
+            fieldValue = field.getValue();
+
+            // update counts
+            counts.put(fieldName, 1 + (counts.containsKey(fieldName) ? counts.get(fieldName) : 0));
+
+            // update running sum
+            fieldSum.put(fieldName, fieldValue + (fieldSum.containsKey(fieldName) ? fieldSum.get(fieldName) : 0));
+
+            // update running deltas
+            delta.put(fieldName, fieldValue * docCount - fieldSum.get(fieldName));
+
+            // update running mean, variance, skewness, kurtosis
+            if (means.containsKey(fieldName) == true) {
```

Comment:
Could we add a reference here? Some paper, book chapter, wikipedia article or something.